src/serde_macros.rs

```diff
@@ -1,86 +1,124 @@
-                            formatter.write_str("a bytestring")
-                        }
-
-                        fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
-                        where
-                            E: $crate::serde::de::Error,
-                        {
-                            $t::from_slice(v).map_err(|_| {
-                                // from_slice only errors on incorrect length
-                                E::invalid_length(v.len(), &stringify!($len))
-                            })
-                        }
-                    }
-
-                    d.deserialize_bytes(BytesVisitor)
-                }
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+        where
+            E: crate::serde::de::Error,
+        {
+            Self::Value::from_hex(v).map_err(E::custom)
+        }
+    }
+
+    struct BytesVisitor<ValueT>(PhantomData<ValueT>);
+
+    impl<'de, ValueT> crate::serde::de::Visitor<'de> for BytesVisitor<ValueT> 
+    where ValueT : SerdeHash {
+        type Value = ValueT;
+
+        fn expecting(&self, formatter: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+            formatter.write_str("a bytestring")
+        }
+
+        fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+        where
+            E: crate::serde::de::Error,
+        {
+            SerdeHash::from_slice_delegated(v).map_err(|_| {
+                // from_slice only errors on incorrect length
+                E::invalid_length(v.len(), &stringify!(N))
+            })
+        }
+    }
+    pub trait SerdeHash
+    where
+        Self: Sized
+            + crate::hex::ToHex
+            + crate::hex::FromHex
+            + std::ops::Index<usize, Output = u8>
+            + std::ops::Index<std::ops::RangeFull, Output = [u8]>
+    {
+        const N: usize;
+
+        fn from_slice_delegated(sl: &[u8]) -> Result<Self, Error>;
+        fn serialize<S: crate::serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+            if s.is_human_readable() {
+                s.serialize_str(&self.to_hex())
+            } else {
+                s.serialize_bytes(&self[..])
             }
         }
-    )
-);
+
+        fn deserialize<'de, D: crate::serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+            if d.is_human_readable() {
+                d.deserialize_str(HexVisitor::<Self>(PhantomData))
+            } else {
+                d.deserialize_bytes(BytesVisitor::<Self>(PhantomData))
+            }
+        }
+    }
+
+    #[macro_export]
+    /// Implements `Serialize` and `Deserialize` for a type `$t` which
+    /// represents a newtype over a byte-slice over length `$len`
+    macro_rules! serde_impl(
+
+        ($t:ident, $len:expr) => (
+            impl $crate::serde_macros::serde_details::SerdeHash for $t {
+                const N : usize = $len;
+                fn from_slice_delegated(sl: &[u8]) -> Result<Self, $crate::Error> {
```

Comment:
N.B. this is done this way to disambiguate which function is being called, and because from_slice is defined variously in Midstate and in HashTrait so the macro is needed to handle that genericity.